File: exercism/haskell/beer-song/Beer.hs
module Beer (sing, verse) where

import Data.List (intercalate)
import Text.Printf (printf)


sing :: Int -> Int -> String
sing a b = intercalate "\n" (map verse [a,a-1..b]) ++ "\n"

template :: String
template = "%d bottles of beer on the wall, %d bottles of beer.\nTake one down and pass it around, %d bottles of beer on the wall.\n"

verse :: Int -> String
verse 0 = "No more bottles of beer on the wall, no more bottles of beer.\nGo to the store and buy some more, 99 bottles of beer on the wall.\n"
verse 1 = "1 bottle of beer on the wall, 1 bottle of beer.\nTake it down and pass it around, no more bottles of beer on the wall.\n"
verse 2 = "2 bottles of beer on the wall, 2 bottles of beer.\nTake one down and pass it around, 1 bottle of beer on the wall.\n"
verse n = printf template n n (n - 1)

File: exercism/csharp/beer-song/Beer.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Beer
{
    static string Template = "{0} bottles of beer on the wall, {0} bottles of beer.\nTake one down and pass it around, {1} bottles of beer on the wall.\n";

    public static string Verse (int n)
    {
        switch (n) {
            case 0:
                return "No more bottles of beer on the wall, no more bottles of beer.\nGo to the store and buy some more, 99 bottles of beer on the wall.\n";
            case 1:
                return "1 bottle of beer on the wall, 1 bottle of beer.\nTake it down and pass it around, no more bottles of beer on the wall.\n";
            case 2:
                return "2 bottles of beer on the wall, 2 bottles of beer.\nTake one down and pass it around, 1 bottle of beer on the wall.\n";
            default:
                return String.Format(Template, n, n - 1);
        }
    }

    public static string Sing (int start, int stop)
    {
        var verseNumbers = Enumerable.Range(stop, start - stop + 1).Reverse();
        return String.Join("\n", verseNumbers.Select(Verse)) + '\n';
    }
}
